crates/phactory/src/prpc_service.rs

```diff
@@ -602,19 +607,23 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> Phactory<Platform>
                 );
             }
             state.recv_mq.dispatch(message);
-        }
 
-        let mut guard = scopeguard::guard(&mut state.recv_mq, |mq| {
-            let n_unhandled = mq.clear();
+            let mut block = BlockInfo {
+                block_number,
+                now_ms,
+                storage: &state.chain_storage,
+                send_mq: &state.send_mq,
+                recv_mq: &mut state.recv_mq,
+                side_task_man: &mut self.side_task_man,
+            };
+
+            system.process_messages(&mut block);
```

Comment:
When moving `process_message` to the inner loop, could it trigger the `on_block` callback for every processed message?

Reply:
Indeed a big problem. Let me try to fix it.

Reply:
Did some refactoring and checked that the first GK's MQ egress matches the old implementation.